File: src/scripts/protocols/ase.ts
import { DataReader } from '../lib/data-reader'
import { InvalidResponseError } from '../lib/response-error'
import { str2bool } from '../lib/str2bool'

interface ASEInfo extends Info {
  gameName: string
  port: string
  serverName: string
  gameType: string
  map: string
  version: string
  private: string
  numPlayers: string
  maxPlayers: string
}

export const info: ProtocolInfo = {
  id: 'ase',
  name: 'All-Seeing Eye',
  feature: 'query',
  transport: 'udp',
}

export const query: QueryFn = () =>
  gsw.send(new Buffer('s'))

const readString = (r: DataReader) => {
  const len = r.u8()

  if (len == 0) {
    throw new InvalidResponseError('string must exist')
  }

  return r.lstring(len - 1)
}

const readGeneralInfo = (r: DataReader): ASEInfo => ({
  gameName: readString(r),
  port: readString(r),
  serverName: readString(r),
  gameType: readString(r),
  map: readString(r),
  version: readString(r),
  private: readString(r),
  numPlayers: readString(r),
  maxPlayers: readString(r),
})

const readKeyValues = (r: DataReader) => {
  const inf: Info = {}

  while (!r.is_end) {
    const key = readString(r)

    if (!key) {
      break;
    }

    const val = readString(r)
    inf[key] = val
  }

  return inf
}

const readPlayerList = (r: DataReader) => {
  let pfields: PlayerField[] = [{
    title: 'Name',
    kind: 'string',
    field: 'name',
    main: true,
  }]

  const players: Player[] = []

  while (!r.is_end) {
    const player: Player = {}
    const flags = r.u8()

    if (flags & 0x01) {
      player.name = readString(r)
    }

    if (flags & 0x02) {
      player.team = readString(r)
    }

    if (flags & 0x04) {
      player.skin = readString(r)
    }

    if (flags & 0x08) {
      player.score = readString(r)
    }

    if (flags & 0x10) {
      player.ping = readString(r)
    }

    if (flags & 0x20) {
      player.time = readString(r)
    }

    players.push(player)
  }

  if (players.length) {
    pfields = Object.keys(players[0]).map(key => ({
      title: key.slice(0, 1).toLocaleUpperCase() + key.slice(1),
      kind: 'string',
      field: key,
      main: key === 'name',
    }))
  }

  return { pfields, players }
}

const normalizeServerInfo = (inf: ASEInfo): ServerInfo => ({
  [InfoField.SERVER_NAME]: inf.serverName,
  [InfoField.GAME_NAME]: inf.gameName,
  [InfoField.GAME_VERSION]: inf.version,
  [InfoField.GAME_MODE]: inf.gameType,
  [InfoField.MAP]: inf.map,
  [InfoField.PRIVATE]: str2bool(inf.private),
  [InfoField.NUM_PLAYERS]: Number(inf.numPlayers),
  [InfoField.MAX_PLAYERS]: Number(inf.maxPlayers),
})

export const processResponse: ProcessResponseFn = data => {
  const r = new DataReader(data)

  const sig = r.lstring(4)

  if (sig !== 'EYE1') {
    throw new InvalidResponseError('not an All-Seeing Eye response')
  }

  const generalInfo = readGeneralInfo(r)
  const keyValues = readKeyValues(r)
  const { pfields, players } = readPlayerList(r)

  const details = { ...generalInfo, ...keyValues }
  const inf = normalizeServerInfo(details)

  gsw.sinfo(details, inf)
  gsw.plist(pfields, players)
}

File: data/protocols/ase.lua
-- Game Server Watcher
-- Copyright (C) 2025  Alexander Burobin

-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Affero General Public License as
-- published by the Free Software Foundation, either version 3 of the
-- License, or (at your option) any later version.

-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU Affero General Public License for more details.

-- You should have received a copy of the GNU Affero General Public License
-- along with this program.  If not, see <https://www.gnu.org/licenses/>.

local DataReader = require "lib/DataReader"
local concat_tables = require "lib/concat_tables"

protocol = {
  id = "ase",
  name = "All-Seeing Eye",
  feature = "query",
  transport = "udp",
}

local info_map = {
  game_name    = "game",
  game_version = "version",
  game_mode    = "game_type",
  server_name  = "name",
  map          = "map",
  num_players  = "num_players",
  max_players  = "max_players",
  private      = "private",
}

function query()
  gsw.send("s")
end

local function read_general_info(r)
  return {
    game_name = r:lstring(),
    port = r:lstring(),
    server_name = r:lstring(),
    game_type = r:lstring(),
    map = r:lstring(),
    version = r:lstring(),
    private = r:lstring(),
    num_players = r:lstring(),
    max_players = r:lstring(),
  }
end

local function read_key_values(r)
  local info = {}

  while not r:is_end() do
    local key = r:lstring()

    if key == "" then
      break
    end

    info[key] = r:lstring()
  end

  return info
end

local function read_player_list(r)
  local players = {}

  while not r:is_end() do
    local player = {}
    local flags = r:u8()

    if (flags & 0x01) > 0 then
      player.name = r:lstring()
    end

    if (flags & 0x02) > 0 then
      player.team = r:lstring()
    end

    if (flags & 0x04) > 0 then
      player.skin = r:lstring()
    end

    if (flags & 0x08) > 0 then
      player.score = r:lstring()
    end

    if (flags & 0x10) > 0 then
      player.ping = r:lstring()
    end

    if (flags & 0x20) > 0 then
      player["time"] = r:lstring()
    end

    players:insert(player)
  end

  return players
end

local function get_pfields(players)
  local pfields = {
    {
      title = 'Name',
      kind = 'string',
      field = 'name',
    },
  }

  if #players > 0 then
    for k, v in pairs(players[0]) do
      pfields:insert({
        title = k:sub(1, 2):upper() .. k:sub(2),
        kind = 'string',
        field = k,
      })
    end
  end

  return pfields
end

local function normalize_server_info(details)
  local info = {}

  for k, v in pairs(info_map) do
    info[k] = details[v]
  end

  return info
end

function process(data)
  local r = DataReader(data)
  local sig = r:string(4)

  if sig ~= "EYE1" then
    error("invalid response: not All-Seeing Eye")
  end

  local general_info = read_general_info(r)
  local values = read_key_values(r)
  local details = concat_tables(general_info, values)
  local inf = normalize_server_info(general_info)
  local players = read_player_list(r)
  local pfields = get_pfields(players)

  gsw.sinfo(details, inf)
  gsw.plist(pfields, players)
end
